projects/remote_transceiver/src/remote_transceiver.cpp

```diff
@@ -62,8 +62,35 @@ HTTPServer::HTTPServer(tcp::socket socket, SailbotDB & db) : socket_(std::move(s
 
 void HTTPServer::doAccept() { readReq(); }
 
-Listener::Listener(bio::io_context & io, tcp::acceptor acceptor, SailbotDB && db)
-: io_(io), acceptor_(std::move(acceptor)), db_(std::move(db)){};
+Listener::Listener(bio::io_context & io, tcp::endpoint endpoint, SailbotDB && db)
+: io_(io), acceptor_(bio::make_strand(io)), db_(std::move(db))
+{
+    beast::error_code ec;
+
+    try {
+        acceptor_.open(endpoint.protocol(), ec);
+        if (ec) {
+            throw(ec);
+        }
+
+        acceptor_.set_option(bio::socket_base::reuse_address(true), ec);
+        if (ec) {
+            throw(ec);
+        }
+
+        acceptor_.bind(endpoint, ec);
+        if (ec) {
+            throw(ec);
+        }
+
+        acceptor_.listen(bio::socket_base::max_listen_connections, ec);
+        if (ec) {
+            throw(ec);
+        }
+    } catch (beast::error_code ec) {
+        std::cerr << "Error: " << ec.message() << std::endl;
+    }
+};
 
 void Listener::run()
```

Comment:
in the official example, the equivalent of our `run()` utilizes `beast::bind_front_handler()` and passes it a pointer to a helper function which calls `async_accept`. When I try this the tests hang, do you have any idea why?

Reply:
The general concept of a `bind()` function is to call a callback function respective to a specific instance of an object. In the example, `bind()` is used to associate the callback function `on_accept()` with the current instance of the `listener` class. Currently in remote_transceiver, a lambda function is used to accomplish the same thing (mostly). I personally like [this explanation of lambdas](https://learn.microsoft.com/en-us/cpp/cpp/lambda-expressions-in-cpp?view=msvc-170). In short:

1. Specifying `[&]` in the lambda means that it's capturing all variables in the current scope.
2. This means we are capturing the current instance of the `Listener` object and its member variables/functions.
3. This accomplishes the same thing as `bind()`, but with less indirection. The callback function is super short and only used by this one `acceptor_.async_accept()` line, so why write a new function somewhere else in the source file?

tl;dr they do the same thing.

Reply:
To be 100% equivalent, you can change the `[&]` to `[this]` or `[shared_from_this()]` like in `readReq()` [here](https://github.com/UBCSailbot/network_systems/blob/ed2b1d5529d482227ad230f5821aa0b054c3a2c8/projects/remote_transceiver/src/remote_transceiver.cpp#L87-L88). I didn't realize I captured the object instance in `run()` and `readReq()`, so please change them to whatever you like so long as it's consistent.

Reply:
changed listener to use `this` instead of `&`. Tried using shared_from_this() but it doesn't let me. I'm thinking it may be because Listener is created using `std::unique_ptr` instead of `shared_ptr()`